apps/settings/js/modules/bluetooth/bluetooth_context.js

```diff
@@ -0,0 +1,852 @@
+      }
+
+      this.numberOfPairedDevices = pairedDevices.length;
+    },
+
+    /**
+     * Since we have three properties which are relative with hardware status,
+     * update them via this method here.
+     * Update status, enabled, settings key 'bluetooth.enabled'
+     * The input state would be {'disabled', 'disabling', 'enabled', 'enabling'}
+     *
+     * @access private
+     * @memberOf BluetoothContext
+     * @param {String} state
+     */
+    _updateStatus: function btc__updateStatus(state) {
+      // Wrapper state to enabled/disabled toggle state.
+      var enabled;
+      if (state === 'enabled' || state === 'disabling') {
+        enabled = true;
+      } else if (state === 'disabled' || state === 'enabling') {
+        enabled = false;
+      }
+
+      // Sync with settings key
+      SettingsCache.getSettings((results) => {
+        var btEnabled = results['bluetooth.enabled'];
+        if (btEnabled !== enabled) {
+          settings.createLock().set({'bluetooth.enabled': enabled});
+        }
+
+        // Update state
+        this.state = state;
+        this.debug('_updateStatus(): set state = ' + state);
+
+        // Update enabled
+        this.enabled = enabled;
+        this.debug('_updateStatus(): set enabled = ' + enabled);
+      });
+    },
+
+    /**
+     * Set Bluetooth enable/disable.
+     *
+     * @access public
+     * @memberOf BluetoothContext
+     * @param {Boolean} enabled
+     */
+    setEnabled: function btc_setEnabled(enabled) {
+      this.debug('setEnabled(): request set enabled = ' + enabled);
+      if ((this.enabled === enabled) ||
+          (this.state === 'enabling') ||
+          (this.state === 'disabling')) {
+        return;
+      }
+
+      if (!this._defaultAdapter) {
+        return;
+      }
+
+      if (enabled) {
+        this._defaultAdapter.enable().then(() => {
+          this.debug('setEnabled(): set enable successfully :)');
+          this._updateStatus('enabled');
+        }, (reason) => {
+          this.debug('setEnabled(): set enable failed: reason = ' + reason);
+        });
+      } else {
+        this._defaultAdapter.disable().then(() => {
+          this.debug('setEnabled(): set disable successfully :)');
+          this._updateStatus('disabled');
+        }, (reason) => {
+          this.debug('setEnabled(): set disable failed: reason = ' + reason);
+        });
+      }
+    },
+
+    /**
+     * Set Bluetooth discoverable.
+     *
+     * @access public
+     * @memberOf BluetoothContext
+     * @param {Boolean} enabled
+     */
+    setDiscoverable: function btc_setDiscoverable(enabled) {
+      if ((this.discoverable === enabled) || (this.state !== 'enabled')) {
+        return;
+      }
+
+      if (!this._defaultAdapter) {
+        return;
+      }
+
+      this._defaultAdapter.setDiscoverable(enabled).then(() => {
+        this.debug('setDiscoverable(): set discoverable successfully :)');
+        this.discoverable = enabled;
```

Comment:
We don't have to update the property here as we should wait for the attribute change event.